app/src/main/java/com/naxtlevelofandroiddevelopment/marvelgallery/network/providers/Provider.kt

```diff
@@ -0,0 +1,10 @@
+package com.naxtlevelofandroiddevelopment.marvelgallery.network.providers
+
+abstract class Provider<T>(val initializer: () -> T) {
+
+    var original: T? = null
+    var override: T? = null
+
+    fun get(): T = override ?: original ?: initializer().apply { original = this }
```

Comment:
It is not clear? use property override, if null then original, if null then result of expression